File: 11015/Main.java
import java.io.*;

public class Main {
	final static int MAX_DISTANCE = 1000;
	public static void main(String args[]) throws IOException {
		int testCase = 1;
		String line;
		StringBuilder answer = new StringBuilder();
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

		while ((line = in.readLine()) != null) {
			String[] sp = line.split(" ");
			int n = Integer.parseInt(sp[0]);
			int m = Integer.parseInt(sp[1]);
			if (n == 0)
				break;
			int[][] distance = new int[n][n];
			String[] name = new String[n];
			for (int i = 0; i < n; ++i) {
				name[i] = in.readLine();
				for (int j = 0; j < n; ++j) {
					distance[i][j] = (i == j) ? 0 : MAX_DISTANCE;
				}
			}
			while (m-- > 0) {
				line = in.readLine();
				sp = line.split(" ");
				int i = Integer.parseInt(sp[0]) - 1;
				int j = Integer.parseInt(sp[1]) - 1;
				int k = Integer.parseInt(sp[2]);
				distance[i][j] = distance[j][i] = k;
			}
			for (int k = 0; k < n; ++k) {
				for (int i = 0; i < n; ++i) {
					for (int j = 0; j < n; ++j) {
						if (distance[i][j] > (distance[i][k] + distance[k][j])) {
							distance[i][j] = distance[i][k] + distance[k][j];
						}
					}
				}
			}
			int min = Integer.MAX_VALUE;
			int index = 0;
			for (int i = 0; i < n; ++i) {
				int sum = 0;
				for (int j = 0; j < n; ++j) {
					sum += distance[i][j];
				}
				if (sum < min) {
					index = i;
					min = sum;
				}
			}
			answer.append("Case #");
			answer.append(Integer.toString(testCase));
			answer.append(" : ");
			answer.append(name[index]);
			answer.append("\n");
			++testCase;
		}
		System.out.print(answer);
	}
}

File: 11015/main.py
from sys import stdin

def main():
	case = 1
	while True:
		n, m = map(lambda x: int(x), stdin.readline().split())
		if n == 0:
			break
		names = []
		distance = [[1000 for j in range(n)] for i in range(n)]
		for x in range(n):
			distance[x][x] = 0
		for x in range(n):
			names.append(stdin.readline().strip())
		for x in range(m):
			i, j, k = map(lambda x: int(x), stdin.readline().split())
			i -= 1
			j -= 1
			distance[i][j] = distance[j][i] = k
		for k in range(n):
			for i in range(n):
				for j in range(n):
					if distance[i][j] > (distance[i][k] + distance[k][j]):
						distance[i][j] = distance[i][k] + distance[k][j]
		min = 100000
		name = names[0]
		for i in range(n):
			s = sum(distance[i])
			if s < min:
				min = s
				name = names[i]
		print("Case #%d : %s" % (case, name))
		case += 1

if __name__ == '__main__':
	main()
